src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -491,9 +497,83 @@ public static String getIndexName(String cfName)
         return StringUtils.substringAfter(cfName, Directories.SECONDARY_INDEX_NAME_SEPARATOR);
     }
 
+    // TODO: JavaDoc if this approach survives...mention that this should not be used for live SSTables
+    public void validateStorageAttachedIndexes(Collection<SSTableReader> sstables) throws IOException
+    {
+        // TODO: A more generic approach via the Index interface rather than just working w/ SAI 
+        Set<StorageAttachedIndex> toValidate = indexes.values().stream().filter(i -> i instanceof StorageAttachedIndex)
+                                                                        .map(i -> (StorageAttachedIndex) i)
+                                                                        .collect(Collectors.toSet());
+
+        if (toValidate.isEmpty())
+            return;
+
+        for (SSTableReader sstable : sstables)
+        {
+            IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+
+            if (!indexDescriptor.isPerSSTableIndexBuildComplete())
+                throw new IllegalStateException(indexDescriptor.logMessage("Incomplete per-SSTable index"));
+
+            indexDescriptor.checksumPerSSTableComponents();
+
+            for (StorageAttachedIndex index : toValidate)
+                indexDescriptor.checksumPerIndexComponents(index.getIndexContext());
```

Comment:
TODO: Might want to check the existence of the per-column components here to avoid any useless validation...